rusty_cards/server/src/server.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    io::{self, Read, Write},
+    net::{Shutdown, SocketAddr, TcpListener, TcpStream},
+    sync::{atomic::AtomicBool, Mutex},
+};
+
+use rusty_cards::Handshake;
+
+// Calss that sotres the stream of the client and
+// socket address that the client sent to the server
+struct ClientsAddr {
+    stream: TcpStream,
+    listener_addr: SocketAddr,
+}
+
+impl ClientsAddr {
+    fn new(stream: TcpStream, listener_addr: SocketAddr) -> Self {
+        ClientsAddr {
+            stream,
+            listener_addr,
+        }
+    }
+
+    fn get_stream(&self) -> &TcpStream {
+        &self.stream
+    }
+
+    fn get_listener_addr(&self) -> SocketAddr {
+        self.listener_addr
+    }
+}
+
+pub struct Server {
+    tcp_listener: TcpListener,
+    running: AtomicBool,
+}
+
+impl Default for Server {
+    fn default() -> Self {
+        println!("Binding server to 127.0.0.1:8888");
+        let tcp_listener = TcpListener::bind("127.0.0.1:8888").unwrap();
+        Server {
+            tcp_listener,
+            running: AtomicBool::new(true),
+        }
+    }
+}
+
+impl Server {
+    pub fn start(&self) {
+        println!("Starting the server on port {}...", self.port());
+        let players = Mutex::new(Vec::new());
+        std::thread::scope(|s| {
+            for stream in self.tcp_listener.incoming() {
+                println!("Client appeared!");
+                let stream = stream.unwrap();
+                if !self.running.load(std::sync::atomic::Ordering::Relaxed) {
+                    break;
+                }
+
+                s.spawn(|| {
+                    match Self::handle_connection(&mut players.lock().unwrap(), stream) {
```

Comment:
to niedobrze dawać wnętrze zablokowanego muteksa do funkcji, bo blokuje to inne wątki

Reply:
to jest bardzo poważny błąd z dziedziny programowania współbieżnego